### DISCUSSION TITLE
Неудобно работать с клавиатурой в ModalPage на iOS

### DISCUSSION THREAD

**Discussion by shevchux** [VKCOM/VKUI#3792] [Ideas]
**Вводная**

Важная iOS особенность, которую нужно учитывать в библиотеке: при подъеме клавиатуры вьюпорт не уменьшается, а уезжает за пределы экрана. Как следствие, при поднятой клавиатуре образуется дополнительный скролл, который позволяет проскроллить так называемый visualViewport.

![image](https://user-images.githubusercontent.com/22776192/207424321-89737dd1-1797-4e22-b14b-4d455b23f3ed.png)

**Описание проблемы**

Неудобно работать с клавиатурой в ModalPage на iOS, если внутри ModalPage есть много скролящегося контента. Конфликтуют кастомный скролл внутри ModalPage и скролл iOS visualViewport. Высокий шанс свернуть ModalPage при попытке скролльнуть вверх.

**Окружение**
Любой iOS. Любой VKUI.
Воспроизводится в мини-приложениях как в нативном клиенте, так и в браузере (Chrome, Safari).

**Пример с кодом**
https://codesandbox.io/s/zealous-currying-lgcyss?file=/src/index.js

**Видео**

<details>
<summary>Демо</summary>

https://user-images.githubusercontent.com/22776192/207424905-858b6dac-1c62-4b22-abb0-f672b4cde03c.MP4
</details>

2-3 сек: Делаем скролл вниз. Не удается доскролить до низа внутреннего скролла ModalPage. Так как часть вьюпорта под клавиатурой.
3-4 сек: Необходимо сделать паузу, чтобы взять управление над скроллом visualViewport (Проблема 1. Неудобно)
5-7 сек: Делаем скролл вверх. Не удается доскролить до верха скролла ModalPage. Так как часть вьюпорта выше физического экрана.
8-9 сек: Необходимо сделать паузу, чтобы взять управление над скроллом visualViewport (Проблема 2. Случайно закрываем модалку, не получаем желаемый результат. Решение Мастхэв.)

**Предлагаемое решение**

- Использовать visualViewport API. При поднятой клавиатуре скролл-боксу внутри ModalPage добавить нижний отступ, чтобы скролл на шаге видео "2-3 сек" работал сразу. Опираться на vkBridge инсет (который шлет только клиент ВК) не предлагаю, так как решение напрашивается и для браузеров.
- Так же после шага видео "2-3 сек" нужно запретить скролл visualViewport, чтобы избежать проблем после шага видео "3-4 сек" (остальные автоматом будут не актуальны). Напрямую это не сделать никак, только через превент touchMove (если мы уже в нижнем положении скролла ModalPage – превентить touchMove вниз, чтобы не брать управление над скроллом visualViewport)

**Альтернативное простое решение**

Запретить при поднятой клавиатуре закрывать модалку жестом вниз. Проблемы с необходимостью делать паузу останутся, но хотя бы решится мастхэв проблема 2, предотвратим случайное сворачивание.
Идея пересекается с запросом [https://github.com/VKCOM/VKUI/issues/3628](https://vk.com/away.php?to=https%3A%2F%2Fgithub.com%2FVKCOM%2FVKUI%2Fissues%2F3628&cc_key=).

**Test Cases**

1. (!!!) Учесть случай, когда поле ввода находится не вверху модалки, но и в её середине или внизу. В этом случае iPhone постарается сделать перемещение visualViewport так, как он сам пожелает оптимальным. С этим надо как-то уживаться, возможно еще решение выше не полностью продумано.
2. Помнить, что высота клавиатуры может меняться, например, когда переключаешься между вводом текста и emoji
3. Не помешала бы корректная работа FixedLayout vertical="bottom" внутри модалки. В этом случае понадобится переопределить safe-area-inset-bottom на величину клавиатуры. Но не глобально, а в области стилей ModalPage.

**Comment by shevchux** [comment]
Test Case 1. В интернете описывают лайфхаки, как предотвратить автоматическое перемещение visualViewport: перед фокусировкой надо на долю секунды перенести поле ввода очень высоко (transform) и вернуть обратно. Кстати, на этом приёме работает мессенджер внутри VK Знакомств (обратите внимание, как интерфейс вписывается под клавиатуру, при этом visualViewport не уезжает, шапка продолжает быть видна)

**Comment by k-egor-smirnov** [comment]
Как раз для таких целей делали библиотеку https://github.com/VKCOM/IOSDevice
Возможно, стоит ее допилить и написать пример использования в VKUI

**Reply by shevchux** [reply]
Мне еще понравился такой источник. Можно подцепить чего.
https://github.com/adamjgrant/a-viewport-that-works

**Comment by inomdzhon** [comment]
> Использовать visualViewport API. При поднятой клавиатуре скролл-боксу внутри ModalPage добавить нижний отступ, чтобы скролл на шаге видео "2-3 сек" работал сразу.

Пока не совсем понял как это должно работать 🤔

**Reply by shevchux** [reply]
Предлагалось узнавать высоту виртуальной клавиатуры через бридж или visualViewport API. На основе этого знания делать отступ скролл-бокса внутри ModalPage при поднятой клавиатуре. Так, чтобы весь бокс был в видимой части. А белый отступ был под виртуальной клавиатурой.

![image](https://user-images.githubusercontent.com/22776192/208838404-1dd45591-b689-429f-99bd-2470e666c030.png)

**Reply by shevchux** [reply]
Такое решение будет сегодня вечером в проде vk.com/dating в фильтрах, в выборе города.

**Reply by inomdzhon** [reply]
> Такое решение будет сегодня вечером в проде vk.com/dating в фильтрах, в выборе города.

Сможешь плиз маякнуть когда зарелизити?

**Reply by shevchux** [reply]
Можно смотреть. Фильтры, выбор города.

https://user-images.githubusercontent.com/22776192/208903636-fc3e4de5-a926-4f03-9f2d-7d3ea0aa9260.MOV

**Reply by shevchux** [reply]
Только на бридже, только внутри клиента пока

**Reply by inomdzhon** [reply]
Ты `.vkuiModalPage__content` задаёшь `margin-bottom` или как?

**Reply by shevchux** [reply]
```margin-bottom``` у ```vkuiModalPage__content-wrap```. Зануляю ```padding-bottom``` (равен ```--safe-area-inset-bottom```) у ```ModalPage__content-in```. Переопределяю  ```--safe-area-inset-bottom``` у ```ModalPage__content-in``` высотой клавиатуры. Таким образом, если понадобится использовать FixedLayout внутри, оно норм приживётся.


Хук с фиксом. Хук возвращает реф. Реф вешаешь на проп ```getModalContentRef``` у ```ModalPage```.

<details><summary>Details</summary>
<p>

```tsx
import { useEffect, useRef } from 'react';
import { useIPhoneKeyboardState } from 'src/hooks/useIPhoneKeyboardState';

export function useModalPageKeyboardSafe() {
  const modalContentRef = useRef<HTMLDivElement>(null);
  const iOSKeyboardState = useIPhoneKeyboardState();

  useEffect(() => {
    const modalContentEl = modalContentRef.current;
    const modalContentInEl = modalContentEl?.firstElementChild as HTMLDivElement | undefined;
    const modalContentWrapEl = modalContentEl?.parentElement;

    if (!modalContentEl || !modalContentInEl || !modalContentWrapEl) {
      return;
    }

    if (iOSKeyboardState.supports && iOSKeyboardState.opened) {
      modalContentWrapEl.style.marginBottom = `${iOSKeyboardState.height}px`;
      modalContentWrapEl.style.transition = `margin-bottom 0.5s ease`;
      modalContentInEl.style.paddingBottom = '0';
      modalContentInEl.style.setProperty('--safe-area-inset-bottom', `${iOSKeyboardState.height}px`);

      let startTouchClientY = 0;

      const touchStartListener = (e: TouchEvent) => {
        startTouchClientY = e.changedTouches[0].clientY;
      };

      const touchMoveListener = (e: TouchEvent) => {
        const { scrollHeight, scrollTop, offsetHeight } = modalContentEl;
        const isScrolledToBottom = offsetHeight + scrollTop >= scrollHeight;
        const hasNotScroll = offsetHeight >= scrollHeight;
        const isScrolledToTop = scrollTop <= 0;

        const shiftY = e.changedTouches[0].clientY - startTouchClientY;
        const isMoveToBottom = shiftY < 0;
        const isMoveToTop = shiftY > 0;

        if (hasNotScroll || (isMoveToBottom && isScrolledToBottom) || (isMoveToTop && isScrolledToTop)) {
          e.preventDefault();
        }
      };

      modalContentWrapEl.addEventListener('touchmove', touchMoveListener, { capture: true });
      modalContentWrapEl.addEventListener('touchstart', touchStartListener, { capture: true });

      return () => {
        modalContentWrapEl.removeEventListener('touchmove', touchMoveListener, { capture: true });
        modalContentWrapEl.removeEventListener('touchstart', touchStartListener, { capture: true });
      };
    }

    modalContentWrapEl.style.transition = '';
    modalContentWrapEl.style.marginBottom = '';
    modalContentInEl.style.paddingBottom = '';
    modalContentInEl.style.removeProperty('--safe-area-inset-bottom');
  }, [iOSKeyboardState]);

  return modalContentRef;
}
```

</p>
</details> 

useIphoneKeyboardState – хук с интерфейсом ```{ supports: false } | { supports: true; height: number; opened: boolean }```.


_______

Этот способ в текущей реализации будет работать только если поле ввода расположено вверху полно-раскрытой модалки. Если поле ввода будет где-то внизу, нативный iOS сделает автоподстролл visibleViewport туда, куда он посчитает нужным. С этим бороться еще сложнее. Кажется лучше добиться автозакрытия клавиатуры при touchStart на внешнюю область + запретить сворачивание модалки при открытой клавиатуре

**Comment by inomdzhon** [comment]
А что если мы будем скрывать клавиатуру если пользователь "увёл" фокус с поля? Например, как в системе iOS или в Yandex.Go

**Reply by shevchux** [reply]
Сразу touchstart любой области кроме поля ввода и клавиатуры? Гений. Нравится. Легкое решение, которого должно быть достаточно, чем делать довольно сложные понты. Главное это более нативно. ВК в любых местах, где есть поиск, делает аналогично.

**Reply by inomdzhon** [reply]
Угу

<details>
<summary>Демо-видео</summary>

https://user-images.githubusercontent.com/5850354/208901011-b4ac560b-5cfd-4bfd-bd37-e97d58a2309e.MP4
</details>

Код из **Демо-видео** по ссылке https://codesandbox.io/s/vkui-discussions-3792-bottom-sheet-crlmv

## ⚠️ Фиксированная шапка с полем ввода

>  см. 00:15 минуту в **Демо-видео**

1. Делаем фокус на поле | _<клавиатура показалась>_
3. Листаём контент, чтобы он продолжал прокручиваться | _<клавиатура скрылась>_
5. Докручиваем контент, который уже был процессе прокрутки
7. Делаем фокус на поле | _<клавиатура показалась>_
8. Продолжаем докручивать контент | ⚠️ _<клавиатура НЕ скрывается>_

**Итог:** пока скрол не остановится `touchstart` не сработает

**Reply by shevchux** [reply]
При поднятой виртуальной клавиатуре создать создать слой, который будет ловить клик?

**Reply by inomdzhon** [reply]
> При поднятой виртуальной клавиатуре создать создать слой, который будет ловить клик?

Вариант, но тогда нужно делать его из нескольких частей, чтобы само поле не перекрывалось этим слоем

---

То, что `touchstart` не вызывается при действующем скролле извеcтная проблема оказывается https://developer.apple.com/forums/thread/99316
